### DISCUSSION TITLE
optimization features - pausing scene

### DISCUSSION THREAD

**Discussion by dennemark** [brianzinn/react-babylonjs#113] [Ideas]
Hi,
in my latest pull request I have extended the snippet material example with code to freeze a node material for better performance and unfreeze it as soon as its values are being updated. This might be extendable to other materials. But I was rather wondering, if we could adapt this for other purposes such as i.e. rendering. Recently I saw this webgl-library, which is pausing rendering if nothing is happening: https://github.com/hmans/three-elements
Since we should know about the states of our components, we could only trigger rendering, if we interact with the scene. Actually doing this for camera movement might be manageable, as soon as the camera moves, the render loop continues. However, if we have elements that change color i.e. on mouse over and the camera is not moving, the color might not get updated. Therefore, it would require maybe a hook that would fire a rerender of the scene. Or if all events within babylon are observable, we could trigger rerender. However, I guess it is not so easy.
Eitherway, I could imagine adding a pause parameter to the Scene component might be a good addition? https://forum.babylonjs.com/t/solved-possibility-to-pause-rendering-without-removing-render-loops/5921/3

**Comment by brianzinn** [comment]
That repo came up in my github feed today - looks cool!  I was thinking about it a little bit.  I think it's easy enough to stop the render loop like in that thread.  To have it on auto will need more than pointer events.  Sometimes the camera is moving (maybe we can listen for that), but also there could be a mesh moved or rotated, but then I suppose we could put a `pause` here and leave that up to the developer:
https://github.com/brianzinn/react-babylonjs/blob/master/src/Engine.tsx#L72

Might be good to have as an enum to allow more options.  I'm thinking 'always', 'never', 'minimum', 'alternate' or something.  Maybe we don't need that many, but minimum would render on pointer down until pointer up (or left screen) and monitor cameras.  Maybe 'alternate' could be used to only render some frames - like every other frame for low power devices.  Anyway, it's an interesting idea and it caught my attention as well.

**Comment by dennemark** [comment]
Sounds like a good starting point! I only looked over three-elements code shortly, but I had the feeling that it listens to all possible events and then fires the render, something like an event pipeline close to the render loop.  Maybe the enum can be some kind of profiles, too. Similar to the Scene Optimization features of babylon with preconfiguration or you do it on your own (https://doc.babylonjs.com/divingDeeper/scene/sceneOptimizer). 
Having pointer down and up events would not allow hover anymore right? If pointer events can be tracked without using the render loop, then hovering could still work. Otherwise the render would run as long as the mouse is being moved over the canvas - touch devices would have an advantage here.

**Comment by dennemark** [comment]
Hey,
just tried out another little optimization: hardware scaling on movement. This event is fired on scene.onPointerObservable. 
(Since my current react-babylonjs version does not integrate this on JSX as i.e. onPointerMove, the scene object is missing if I use JSX. So I just added the observer via useEffect.)

```
  const onPointerInteract = (pointer: PointerInfo) =>{
    if(scene !== null){
      let pointerEvent = pointer.event as PointerEvent;  
      let scaling = 1.0;
      if(pointer.type === PointerEventTypes.POINTERWHEEL  // on scroll
        || (pointer.type === PointerEventTypes.POINTERMOVE // on movement
            && pointerEvent.pressure > 0)){
          scaling = 4.0;
      }
      
      scene.getEngine().setHardwareScalingLevel(scaling);
    }
  }
```

**Comment by dennemark** [comment]
This thread has nice discussion and playgrounds on scene optimization:
https://forum.babylonjs.com/t/what-is-the-correct-way-to-render-on-demand/17256/22
Instead of listening to pointer events, it only renders on camera movement. It needs a custom render loop.

I have created a hook. However, it is not well tested yet and it is missing trigger of scene.render() on i.e. loading of assets. So you might only see your assets after moving the camera.

```
import {useEffect, useState} from 'react'
import { Observer } from "@babylonjs/core/Misc/observable"
import { PointerInfo, PointerEventTypes } from "@babylonjs/core/Events/pointerEvents"
import { Scene } from "@babylonjs/core/scene"
import {useScene} from "react-babylonjs"
import { Camera, ArcRotateCamera, TargetCamera } from '@babylonjs/core/Cameras'

let first = true;
let hardwareScalingFallback = 1;
const isArcRotateCameraStopped = (camera: ArcRotateCamera) => {
    
    return camera.inertialAlphaOffset === 0 
        && camera.inertialBetaOffset === 0
        && camera.inertialRadiusOffset === 0
        && camera.inertialPanningX === 0
        && camera.inertialPanningY === 0;
};

const isTargetCameraStopped = (camera: TargetCamera) => {
    return camera.cameraDirection.x === 0 
        && camera.cameraDirection.y === 0
        && camera.cameraDirection.z === 0
        && camera.cameraRotation.x === 0
        && camera.cameraRotation.y === 0;
};

const isCameraStopped = (camera: Camera) => {
    return camera instanceof ArcRotateCamera 
        ? isArcRotateCameraStopped(camera as ArcRotateCamera) 
        : camera instanceof TargetCamera 
        ? isTargetCameraStopped(camera as TargetCamera)
        : false;
}

const useRenderOnDemand = (scene: Scene | null, hardwareScaling?: number) => {
    const [on, setOn] = useState(false);

    const [camera, setCamera] = useState<Camera | undefined>(undefined);

    useEffect(()=>{
        if(scene){
            scene.activeCamera && setCamera(scene.activeCamera)
            const updateCamera = scene.onActiveCameraChanged.add(() => {scene.activeCamera && setCamera(scene.activeCamera)})
            return ()=>{
                scene.onActiveCameraChanged.remove(updateCamera)
            }
        }
    },[scene])

    useEffect(()=>{

        if(scene && camera){
            const engine = scene.getEngine();

            hardwareScalingFallback = engine.getHardwareScalingLevel()
            
            var renderLoop = function () {
                //camera.update();
                if (!isCameraStopped(camera)) {
                    hardwareScaling && engine.setHardwareScalingLevel(hardwareScaling);
                    scene.render(true, true)
                    first = true;
                } else {
                    if (first) {
                        hardwareScaling && engine.setHardwareScalingLevel(hardwareScalingFallback);
                        scene.render(true, true)
                        first = false;
                    }
                }
            };
            engine.stopRenderLoop();
            engine.runRenderLoop(renderLoop);
            return () => {
                engine.stopRenderLoop();
                engine.runRenderLoop(()=>{scene.render()});
            }
        }
    },[scene, camera])
    return on;
} 

export default useRenderOnDemand;
```

**Reply by brianzinn** [reply]
That's great.  We could perhaps add a way for all of the useAsset to set a state when their task completes (ie: onMeshLoaded).  It would perform really well on lower devices for simple scenes (ie: meshes aren't moving in scene, etc.)

**Reply by dennemark** [reply]
Maybe you have followed this discussion on twitter: https://twitter.com/0xca0a/status/1417466656339681280
r3f checks for changes in all props and invalidates the renderer allowing to call a render on demand.
I am not familar with react-reconciler and do not know, if it is difficult to implement.

Impementing a rerender within useAsset would be ok, too, I guess. But most probably there will be other components that might ask for rerender on prop change, too. So each one would need to know about a function to render the scene.

Within the discussion Paul was pointing out, that my hook could be named more accurate, too ;) Something like useRenderOnCameraMovement would be more accurate.

**Reply by brianzinn** [reply]
Interesting discussion.  I feel like I want to split this out into a new project where lots of utilities like these hooks can be added.

For loading assets there is a scene observer for when meshes are added that might be useful.  Also, cameras have observers on when their projection matrix changes that may be useful as opposed to looking at different camera properties and epsilons.

**Reply by brianzinn** [reply]
more interest in scene render optimization:
https://github.com/brianzinn/react-babylonjs/issues/181
